java/arcs/core/util/Parser.kt

```diff
@@ -180,17 +202,44 @@ sealed class ParseResult<out T>() {
     data class Success<out T>(
         val value: T,
         override val start: SourcePosition,
-        override val end: SourcePosition
+        override val end: SourcePosition,
+        override val consumed: Int = 1
     ) : ParseResult<T>()
 
     /** Represents a parse failure. */
-    class Failure(
+    data class Failure(
         val error: String,
         override val start: SourcePosition,
-        override val end: SourcePosition
-    ) : ParseResult<Nothing>()
+        override val end: SourcePosition,
+        override val consumed: Int = 0,
+        val parser: String = "",
+        val cause: Failure? = null
+    ) : ParseResult<Nothing>() {
+        override fun toString() = rootCause(this).let {
+            "${it.error} at line ${it.start.line}, column ${it.start.column}"
+        } + (this.cause?.let { "\n[Traceback]" + traceBack(this) })
+    }
+}
+
+private tailrec fun rootCause(cause: Failure): Failure =
+    if (cause.cause == null) cause else rootCause(cause.cause)
+
+private fun traceBack(cause: Failure?, indent: Int = 1): String = when {
+    cause == null -> ""
+    cause.parser.isBlank() -> traceBack(cause.cause, indent)
+    else -> "\n" + ("--".repeat(indent)) + ">${cause.parser}" + traceBack(cause.cause, indent + 1)
```

Comment:
For a complex Paxel expression I think you get more than 10-20 levels deep. With such deep nesting `"--".repeat(indent)` may be an obstacle to readability.

Maybe simply: `else -> "\n> ${cause.parser}" + traceBack(cause.cause)` ?